### ISSUE TITLE
clarify enum grammar in WIT.md

### ISSUE THREAD

**Issue by patrickhuber** [WebAssembly/component-model#132]
I'm trying to understand how optional trailing comma is defined in the enum grammar. 

In the [WIT.md](https://github.com/WebAssembly/component-model/blob/b458b00c73b940a3148cc7a5b3c20af70a487019/design/mvp/WIT.md#item-enum-variant-but-with-no-payload), enums are defined as follows:

```
enum-items ::= 'enum' id '{' enum-cases '}'

enum-cases ::= id,                     // Is this comma a typo? Shouldn't it be ','?
             | id ',' enum-cases?
```

Looking at the [wit type tests](https://github.com/bytecodealliance/wasm-tools/blob/001e5fcb1d0bd753e4177c224f7d916e3cdbecbf/crates/wit-parser/tests/ui/types.wit) enums can be defined with the following cases:

```wit
enum t41 { a, b, c }
enum t42 { a, b, c, }
```

It appears that the optional trailing comma is not accounted for in this grammar (or there is a typo). To capture the cases where a trailing comma is optional, shouldn't the definition look something like the following? 

```
enum-items ::= 'enum' id '{' enum-cases '}'

enum-cases ::= id
             | id ','
             | id ',' enum-cases?
```

example with 'enum t41 { a, b, c }'

```
enum-items = 'enum' 't41' '{' enum-cases1 '}'
enum-cases1 = 'a' enum-cases2                         // id ',' enum-cases
enum-cases2 = 'b' enum-cases3                         // id ',' enum-cases
enum-cases3 = 'c'                                                // id
```

example with 'enum t42 { a, b, c, }'

```
enum-items = 'enum' 't41' '{' enum-cases1 '}'
enum-cases1 = 'a' enum-cases2                         // id ',' enum-cases
enum-cases2 = 'b' enum-cases3                         // id ',' enum-cases
enum-cases3 = 'c'                                                // id ','
```

**Comment by patrickhuber** [comment]
Looking at the [variant definition](https://github.com/WebAssembly/component-model/blob/b458b00c73b940a3148cc7a5b3c20af70a487019/design/mvp/WIT.md#item-variant-one-of-a-set-of-types), the same question applies there as well.

**Comment by ia0** [comment]
The non-quoted comma looks like a typo indeed. It should be removed. The same typo is for `union-cases`. The correct version can be seen for `record-fields` for example. Note that the recursive call in the second variant of the definitions is optional and thus the comma would be a trailing comma in that case. The first variant permits to omit the trailing comma. Those definitions implement non-empty comma-separated lists with optional trailing comma.

```
xs ::= x | x ',' xs?
```

**Comment by patrickhuber** [comment]
I see that record-fields is missing the typo, however it still doesn't account for an optional trailing comma. 

In the example you supplied, the following strings are possible (assuming x is something like the string 'x'):

* "x"
* "x,x"
* "x,x,x"

Even with xs being optional, the trailing comma is not possible because xs can not be nil. 

With a grammar like :

`xs ::= x | x ',' | x ',' xs?` the following strings are possible:

* "x"
* "x,"
* "x,x"
* "x,x,"
* "x,x,x"
* "x,x,x,"

**Comment by patrickhuber** [comment]
Oh, I see what you mean by optional. Ignore my previous comment.

**Comment by ia0** [comment]
Yes exactly: `x ',' xs?` has the same semantics as `x ',' | x ',' xs` but with less boilerplate. (#133 looks good to me but I'm in no authority to approve.)

**Comment by lukewagner** [comment]
Agreed.  Thanks a lot for filing, explaining and fixing!
